LightGBM installation error_MacOSX10.14_Cmake

gcc -v
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/c++/4.2.1
Apple LLVM version 10.0.0 (clang-1000.10.44.4)
Target: x86_64-apple-darwin18.2.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

**Issue:**
I followed the installation guide and on 'cmake ..' step got following error:
############

-- The C compiler identification is Clang 4.0.1
-- The CXX compiler identification is Clang 4.0.1
-- Check for working C compiler: /anaconda3/bin/x86_64-apple-darwin13.4.0-clang
-- Check for working C compiler: /anaconda3/bin/x86_64-apple-darwin13.4.0-clang -- broken
CMake Error at /anaconda3/share/cmake-3.12/Modules/CMakeTestCCompiler.cmake:52 (message):
  The C compiler

    "/anaconda3/bin/x86_64-apple-darwin13.4.0-clang"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /Users/adityajuvali/LightGBM/build/CMakeFiles/CMakeTmp
    
    Run Build Command:"/anaconda3/bin/make" "cmTC_4d700/fast"
    /anaconda3/bin/make -f CMakeFiles/cmTC_4d700.dir/build.make CMakeFiles/cmTC_4d700.dir/build
    make[1]: Entering directory '/Users/adityajuvali/LightGBM/build/CMakeFiles/CMakeTmp'
    Building C object CMakeFiles/cmTC_4d700.dir/testCCompiler.c.o
    /anaconda3/bin/x86_64-apple-darwin13.4.0-clang   -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe  -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk   -o CMakeFiles/cmTC_4d700.dir/testCCompiler.c.o   -c /Users/adityajuvali/LightGBM/build/CMakeFiles/CMakeTmp/testCCompiler.c
    Linking C executable cmTC_4d700
    /anaconda3/bin/cmake -E cmake_link_script CMakeFiles/cmTC_4d700.dir/link.txt --verbose=1
    /anaconda3/bin/x86_64-apple-darwin13.4.0-clang -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe  -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names -Wl,-pie -Wl,-headerpad_max_install_names -Wl,-dead_strip_dylibs  CMakeFiles/cmTC_4d700.dir/testCCompiler.c.o  -o cmTC_4d700 
    ld: warning: ignoring file /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/lib/libSystem.tbd, file was built for unsupported file format ( 0x2D 0x2D 0x2D 0x20 0x21 0x74 0x61 0x70 0x69 0x2D 0x74 0x62 0x64 0x2D 0x76 0x33 ) which is not the architecture being linked (x86_64): /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/lib/libSystem.tbd
    ld: dynamic main executables must link with libSystem.dylib for architecture x86_64
    clang-4.0: error: linker command failed with exit code 1 (use -v to see invocation)
    make[1]: *** [CMakeFiles/cmTC_4d700.dir/build.make:87: cmTC_4d700] Error 1
    make[1]: Leaving directory '/Users/adityajuvali/LightGBM/build/CMakeFiles/CMakeTmp'
    make: *** [Makefile:121: cmTC_4d700/fast] Error 2
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:7 (PROJECT)


-- Configuring incomplete, errors occurred!


#######

Please help. I am not able install even other models due to this same error.

> Hi @AdityaJuvali !
> 
> As it can be observed from your logs, CMake cannot find you system Clang compiler. Instead, it uses some compiler from Anaconda installation.
> 
> ```
> -- The C compiler identification is Clang 4.0.1
> -- The CXX compiler identification is Clang 4.0.1
> -- Check for working C compiler: /anaconda3/bin/x86_64-apple-darwin13.4.0-clang
> -- Check for working C compiler: /anaconda3/bin/x86_64-apple-darwin13.4.0-clang -- broken
> ```
> Try to specify right compilers by the following command:
> 
> ```
> export CXX=g++ CC=gcc
> ```
> Or just remove `/anaconda3/bin` from your PATH variable.


Thank you for the supportive response. I tried this and got another error. It's anaconda again. Could you please help on changing the PATH variable that you mentioned. I am a beginner Data science student and haven't handled MacOSX before. 

Error:
-- The C compiler identification is AppleClang 10.0.0.10001044
-- The CXX compiler identification is AppleClang 10.0.0.10001044
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/gcc
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/g++
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at /anaconda3/share/cmake-3.12/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES)
Call Stack (most recent call first):
  /anaconda3/share/cmake-3.12/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /anaconda3/share/cmake-3.12/Modules/FindOpenMP.cmake:466 (find_package_handle_standard_args)
  CMakeLists.txt:78 (find_package)


-- Configuring incomplete, errors occurred!

I removed the variable 'anaconda3/bin' from the PATH. It moved ahead but again gave error as:

CMake Error at /usr/local/Cellar/cmake/3.13.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.13.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.13.1/share/cmake/Modules/FindOpenMP.cmake:473 (find_package_handle_standard_args)
  CMakeLists.txt:78 (find_package)

In my environment (macOS 10.14.1), installation error happens.

```
❯ cmake ..
-- The C compiler identification is AppleClang 10.0.0.10001145
-- The CXX compiler identification is AppleClang 10.0.0.10001145
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at /usr/local/Cellar/cmake/3.13.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.13.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.13.2/share/cmake/Modules/FindOpenMP.cmake:473 (find_package_handle_standard_args)
  CMakeLists.txt:78 (find_package)


-- Configuring incomplete, errors occurred!
See also "/Users/henry/.ghq/github.com/henry0312/LightGBM/build/CMakeFiles/CMakeOutput.log".
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -211,8 +211,8 @@ endif(USE_MPI)
 
 if(USE_OPENMP)
     if(CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang")
-        TARGET_LINK_LIBRARIES(lightgbm ${OpenMP_libomp_LIBRARY})
-        TARGET_LINK_LIBRARIES(_lightgbm ${OpenMP_libomp_LIBRARY})
+        TARGET_LINK_LIBRARIES(lightgbm OpenMP::OpenMP_CXX)
+        TARGET_LINK_LIBRARIES(_lightgbm OpenMP::OpenMP_CXX)
     endif()
 endif(USE_OPENMP)
 

--- a/docs/Installation-Guide.rst
+++ b/docs/Installation-Guide.rst
@@ -146,7 +146,13 @@ Only **Apple Clang** version 8.1 or higher is supported.
 
      git clone --recursive https://github.com/Microsoft/LightGBM ; cd LightGBM
      mkdir build ; cd build
-     cmake ..
+     cmake \
+       -DOpenMP_C_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include" \
+       -DOpenMP_C_LIB_NAMES="omp" \
+       -DOpenMP_CXX_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include" \
+       -DOpenMP_CXX_LIB_NAMES="omp" \
+       -DOpenMP_omp_LIBRARY=/usr/local/opt/libomp/lib/libomp.dylib \
+       ..
      make -j4
 
 gcc

--- a/docs/Installation-Guide.rst
+++ b/docs/Installation-Guide.rst
@@ -149,11 +149,11 @@ Only **Apple Clang** version 8.1 or higher is supported.
 
      # For Mojave (10.14)
      cmake \
-       -DOpenMP_C_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include" \
+       -DOpenMP_C_FLAGS="-Xpreprocessor -fopenmp -I$(brew --prefix libomp)/include" \
        -DOpenMP_C_LIB_NAMES="omp" \
-       -DOpenMP_CXX_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include" \
+       -DOpenMP_CXX_FLAGS="-Xpreprocessor -fopenmp -I$(brew --prefix libomp)/include" \
        -DOpenMP_CXX_LIB_NAMES="omp" \
-       -DOpenMP_omp_LIBRARY=/usr/local/opt/libomp/lib/libomp.dylib \
+       -DOpenMP_omp_LIBRARY=$(brew --prefix libomp)/lib/libomp.dylib \
        ..
 
      # For High Sierra or earlier (<= 10.13)

--- a/docs/Installation-Guide.rst
+++ b/docs/Installation-Guide.rst
@@ -147,7 +147,7 @@ Only **Apple Clang** version 8.1 or higher is supported.
      git clone --recursive https://github.com/Microsoft/LightGBM ; cd LightGBM
      mkdir build ; cd build
 
-     # For Mojave
+     # For Mojave (10.14)
      cmake \
        -DOpenMP_C_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include" \
        -DOpenMP_C_LIB_NAMES="omp" \
@@ -156,7 +156,7 @@ Only **Apple Clang** version 8.1 or higher is supported.
        -DOpenMP_omp_LIBRARY=/usr/local/opt/libomp/lib/libomp.dylib \
        ..
 
-     # For High Sierra or earlier
+     # For High Sierra or earlier (<= 10.13)
      cmake ..
 
      make -j4

--- a/docs/Installation-Guide.rst
+++ b/docs/Installation-Guide.rst
@@ -146,6 +146,8 @@ Only **Apple Clang** version 8.1 or higher is supported.
 
      git clone --recursive https://github.com/Microsoft/LightGBM ; cd LightGBM
      mkdir build ; cd build
+
+     # For Mojave
      cmake \
        -DOpenMP_C_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include" \
        -DOpenMP_C_LIB_NAMES="omp" \
@@ -153,6 +155,10 @@ Only **Apple Clang** version 8.1 or higher is supported.
        -DOpenMP_CXX_LIB_NAMES="omp" \
        -DOpenMP_omp_LIBRARY=/usr/local/opt/libomp/lib/libomp.dylib \
        ..
+
+     # For High Sierra or earlier
+     cmake ..
+
      make -j4
 
 gcc

--- a/docs/Installation-Guide.rst
+++ b/docs/Installation-Guide.rst
@@ -448,7 +448,20 @@ Only **Apple Clang** version 8.1 or higher is supported.
 
      git clone --recursive https://github.com/Microsoft/LightGBM ; cd LightGBM
      mkdir build ; cd build
+
+     # For Mojave (10.14)
+     cmake \
+       -DUSE_MPI=ON \
+       -DOpenMP_C_FLAGS="-Xpreprocessor -fopenmp -I$(brew --prefix libomp)/include" \
+       -DOpenMP_C_LIB_NAMES="omp" \
+       -DOpenMP_CXX_FLAGS="-Xpreprocessor -fopenmp -I$(brew --prefix libomp)/include" \
+       -DOpenMP_CXX_LIB_NAMES="omp" \
+       -DOpenMP_omp_LIBRARY=$(brew --prefix libomp)/lib/libomp.dylib \
+       ..
+
+     # For High Sierra or earlier (<= 10.13)
      cmake -DUSE_MPI=ON ..
+
      make -j4
 
 gcc

--- a/python-package/setup.py
+++ b/python-package/setup.py
@@ -119,14 +119,18 @@ def compile_cpp(use_mingw=False, use_gpu=False, use_mpi=False, nomp=False,
     if nomp:
         cmake_cmd.append("-DUSE_OPENMP=OFF")
     if system() == 'Darwin' and not nomp:
-        # https://github.com/Homebrew/homebrew-core/pull/20589
-        cmake_cmd.extend([
-            '-DOpenMP_C_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include"',
-            '-DOpenMP_C_LIB_NAMES="omp"',
-            '-DOpenMP_CXX_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include"',
-            '-DOpenMP_CXX_LIB_NAMES="omp"',
-            '-DOpenMP_omp_LIBRARY=/usr/local/opt/libomp/lib/libomp.dylib',
-        ])
+        cc = os.environ.get('CC')
+        cxx = os.environ.get('CXX')
+        if not (cc and cc.startswith('gcc') and cxx and cxx.startswith('g++')):
+            # Apple Clang
+            # https://github.com/Homebrew/homebrew-core/pull/20589
+            cmake_cmd.extend([
+                '-DOpenMP_C_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include"',
+                '-DOpenMP_C_LIB_NAMES="omp"',
+                '-DOpenMP_CXX_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include"',
+                '-DOpenMP_CXX_LIB_NAMES="omp"',
+                '-DOpenMP_"omp"_LIBRARY=/usr/local/opt/libomp/lib/libomp.dylib'
+            ])
     if use_hdfs:
         cmake_cmd.append("-DUSE_HDFS=ON")
 

--- a/python-package/setup.py
+++ b/python-package/setup.py
@@ -125,11 +125,11 @@ def compile_cpp(use_mingw=False, use_gpu=False, use_mpi=False, nomp=False,
             # Apple Clang
             # https://github.com/Homebrew/homebrew-core/pull/20589
             cmake_cmd.extend([
-                '-DOpenMP_C_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include"',
-                '-DOpenMP_C_LIB_NAMES="omp"',
-                '-DOpenMP_CXX_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include"',
-                '-DOpenMP_CXX_LIB_NAMES="omp"',
-                '-DOpenMP_"omp"_LIBRARY=/usr/local/opt/libomp/lib/libomp.dylib'
+                '-DOpenMP_C_FLAGS=-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include',
+                '-DOpenMP_C_LIB_NAMES=omp',
+                '-DOpenMP_CXX_FLAGS=-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include',
+                '-DOpenMP_CXX_LIB_NAMES=omp',
+                '-DOpenMP_omp_LIBRARY=/usr/local/opt/libomp/lib/libomp.dylib'
             ])
     if use_hdfs:
         cmake_cmd.append("-DUSE_HDFS=ON")

--- a/python-package/setup.py
+++ b/python-package/setup.py
@@ -118,8 +118,18 @@ def compile_cpp(use_mingw=False, use_gpu=False, use_mpi=False, nomp=False,
         cmake_cmd.append("-DUSE_MPI=ON")
     if nomp:
         cmake_cmd.append("-DUSE_OPENMP=OFF")
+    if system() == 'Darwin' and not nomp:
+        # https://github.com/Homebrew/homebrew-core/pull/20589
+        cmake_cmd.extend([
+            '-DOpenMP_C_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include"',
+            '-DOpenMP_C_LIB_NAMES="omp"',
+            '-DOpenMP_CXX_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include"',
+            '-DOpenMP_CXX_LIB_NAMES="omp"',
+            '-DOpenMP_omp_LIBRARY=/usr/local/opt/libomp/lib/libomp.dylib',
+        ])
     if use_hdfs:
         cmake_cmd.append("-DUSE_HDFS=ON")
+
     if system() in ('Windows', 'Microsoft'):
         if use_mingw:
             if use_mpi:

